File: bin/Data/EditorLua/view/ToolBar.lua
local ToolBar = {}
local Translator = require("EditorLua/core/Translate")
ToolBar.itemSize = 20

function ToolBar:init()
    self.isShow = true
    self.curToolName = "camera"--"move" "rotate" "scale"
    self.selectedColor = 0x77999999
end

function ToolBar:OnSave()
    self.assetMgr:SaveScene(self.app.assetRoot .."/test.scene")
end

function ToolBar:OnUndo()
end

function ToolBar:OnRedo()
end

function ToolBar:OnNewScene()
    self.assetMgr:OpenNewScene()
end

function ToolBar:OnTransformTool(name)
    self.curToolName = name
    self.app.cppApp:setCurTool(name)
end

function ToolBar:DrawTransformTool()
    local toolList = {"camera","move","rotate","scale"}
    local imgList = {}
    imgList[1] = self.assetMgr:getImguiTex("res/img/hand.png")
    imgList[2] = self.assetMgr:getImguiTex("res/img/move.png")
    imgList[3] = self.assetMgr:getImguiTex("res/img/rotate.png")
    imgList[4] = self.assetMgr:getImguiTex("res/img/scale.png")
    local curColor = 0x00ffffff;
    for k,v in ipairs(toolList) do
        if v == self.curToolName then
            curColor = self.selectedColor
        else
            curColor = 0x00ffffff;
        end
        imgui.PushStyleColor(imgui.constant.Col.Button,curColor)
        imgui.PushStyleColor(imgui.constant.Col.ButtonHovered,0x77999999)
        imgui.SameLine()
        if imgui.ImageButton(imgList[k],self.itemSize,self.itemSize) then
           self:OnTransformTool(v)
        end
        if imgui.IsItemHovered(0) then
            imgui.BeginTooltip()
            imgui.SetTooltipText(v)
            imgui.EndTooltip()
        end

        imgui.PopStyleColor(2)
    end
end

function ToolBar:DrawTrigerTool()
    local toolList = {"newScene","save","undo","redo"}
    local handleList = {"OnNewScene","OnSave","OnUndo","OnRedo"}
    local imgList = {}
    imgList[1] = self.assetMgr:getImguiTex("res/img/grid.png")
    imgList[2] = self.assetMgr:getImguiTex("res/img/save.png")
    imgList[3] = self.assetMgr:getImguiTex("res/img/undo.png")
    imgList[4] = self.assetMgr:getImguiTex("res/img/redo.png")
    imgui.PushStyleColor(imgui.constant.Col.Button,0x00ffffff)
    imgui.PushStyleColor(imgui.constant.Col.ButtonHovered,0x77999999)
    for k,v in ipairs(toolList) do
        imgui.SameLine()
        if imgui.ImageButton(imgList[k],self.itemSize,self.itemSize) then
            self[handleList[k]](self)
        end
        if imgui.IsItemHovered(0) then
            imgui.BeginTooltip()
            imgui.SetTooltipText(v)
            imgui.EndTooltip()
        end
    end
    imgui.PopStyleColor(2)
end

function ToolBar:Update()
    if self.isShow == true then
        local winsizeX = imgui.GetWindowSize()
        self.isShow = imgui.BeginChild("ToolBar",winsizeX,self.itemSize+6)
        --self.isShow = imgui.Begin(Translator:translate("ToolBar"),self.isShow)
        self:DrawTrigerTool()
        self:DrawTransformTool()
        imgui.EndChild()
        --imgui.End()
    end
end

function ToolBar:Show(show)
    self.isShow = show
end

ToolBar:init()
return ToolBar

File: Source/UrhoEditor/views/Toolbar.cpp
#include "Toolbar.h"
#include "ctrls/AssetMgr.h"
#include "EditorApp.h"
#include "cmd/CmdMgr.h"

namespace Urho3DEditor
{

Toolbar::Toolbar() 
{
	Init();
}
Toolbar::~Toolbar() 
{
}
void Toolbar::Update() 
{
	auto winSize = ImGui::GetWindowSize();
	ImGui::BeginChild("ToolBar",ImVec2(winSize.x,itemSize+6));
	this->DrawTrigerTool();
	this->DrawTransformTool();
	ImGui::EndChild();
}
void Toolbar::DrawTransformTool() 
{
	Vector<String> toolList = { "camera","move","rotate","scale" };
	Vector<int> imgList;
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/hand.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/move.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/rotate.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/scale.png"));
	
	unsigned curColor = 0x00ffffff;
	for (int i= 0;i< toolList.Size();++i)
	{
		auto item = toolList[i];
		if (item == EditorApp::GetInstance()->GetCurTool()) 
		{
			curColor = selectedColor;
		}
		else
		{
			curColor = 0x00ffffff;
		}
		ImGui::PushStyleColor(ImGuiCol_Button, curColor);
		ImGui::PushStyleColor(ImGuiCol_ButtonHovered, 0x77999999);
		ImGui::SameLine();
		if(ImGui::ImageButton(imgList[i],ImVec2(itemSize,itemSize)))
		{
			this->OnTransformTool(item);
		}
		if(ImGui::IsItemHovered())
		{
			ImGui::BeginTooltip();
			ImGui::SetTooltip(item.CString());
			ImGui::EndTooltip();
		}
		ImGui::PopStyleColor(2);
	}
	
}
void Toolbar::DrawTrigerTool() 
{
	Vector<ToolbarHandle> handleList ={&OnNewScene,&OnSave,&OnUndo,&OnRedo};
	Vector<String> toolList = { "newScene","save","undo","redo" };
	Vector<int> imgList;
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/grid.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/save.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/undo.png"));
	imgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/redo.png"));
	Vector<int> grayImgList;
	grayImgList.Push(0);
	grayImgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/save_gray.png"));
	grayImgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/undo_gray.png"));
	grayImgList.Push(AssetMgr::getInstance()->getImguiTex("res/img/redo_gray.png"));
	Vector<bool> activeList;
	activeList.Push(true);
	activeList.Push(CmdMgr::Instance()->HasNode());
	activeList.Push(CmdMgr::Instance()->CanUnDo());
	activeList.Push(CmdMgr::Instance()->CanReDo());

	ImGui::PushStyleColor(ImGuiCol_Button, 0x00ff0000);
	ImGui::PushStyleColor(ImGuiCol_ButtonHovered, 0x77999999);
	for (int i = 0; i < toolList.Size(); ++i) 
	{
		auto item = toolList[i];
		ImGui::SameLine();
		int tmpImg = imgList[i];
		if(!activeList[i])
		{
			tmpImg = grayImgList[i];
		}
		if(ImGui::ImageButton(tmpImg, ImVec2(itemSize, itemSize)))
		{
			if(activeList[i])
			{
				handleList[i]();
			}
		}
		if (ImGui::IsItemHovered()) {
			ImGui::BeginTooltip();
			ImGui::SetTooltip(item.CString());
			ImGui::EndTooltip();
		}
	}
	ImGui::PopStyleColor(2);
}
void Toolbar::Init() 
{
	//curToolName = "camera";
	EditorApp::GetInstance()->GetCurTool();
   
}
void Toolbar::IntItemSize() 
{ 
	this->itemSize = (int) (this->itemSize * EditorApp::GetInstance()->GetDpiScale() );
}
void Toolbar::OnSave() 
{
	String assetRoot = EditorApp::GetInstance()->GetAssetRoot();
	AssetMgr::getInstance()->SaveScene(assetRoot + "/test.uscene");
	CmdMgr::Instance()->OnSave();
}
void Toolbar::OnRedo() 
{
	CmdMgr::Instance()->ReDo();

}
void Toolbar::OnUndo() 
{
	CmdMgr::Instance()->UnDo();
}
void Toolbar::OnNewScene() 
{
	AssetMgr::getInstance()->OpenNewScene();
}
void Toolbar::OnTransformTool(const String& name)
{
	EditorApp::GetInstance()->SetCurTool(name);
}
}